rust/kernel/io_mem.rs

```diff
@@ -177,6 +177,47 @@ impl<const SIZE: usize> IoMem<SIZE> {
         crate::build_assert!(Self::offset_ok::<T>(offset), "IoMem offset overflow");
     }
 
+    /// Copy memory block from an i/o memory into a buffer.
+    ///
+    /// # Examples
+    /// ```
+    /// # use kernel::prelude::*;
+    /// use kernel::io_mem::{self, IoMem, Resource};
+    ///
+    /// fn test(res: Resource) -> Result {
+    ///     // Create an i/o memory block of at least 100 bytes.
+    ///     let mem = unsafe { IoMem::<100>::try_new(res) }?;
+    ///
+    ///     let mut buffer: [i8; 32] = [0; 32];
+    ///
+    ///     // Memcpy 16 bytes from offset 10 of i/o memory block into the buffer.
+    ///     mem.try_memcpy_fromio(&mut buffer, 10, 16)?;
+    ///
+    ///     Ok(())
+    /// }
+    /// ```
+    pub fn try_memcpy_fromio(&self, from: &mut [i8], offset: usize, count: usize) -> Result {
```

Comment:
Maybe call the argument `to` or `buf` instead of `from`? At first I thought it read from the `from: &mut [i8]` and was thinking why it used `&mut` instead of `&`. Also `u8` would be nicer than `i8` I think. `u8` is the "default" type for a single byte.

Reply:
> Maybe call the argument `to` or `buf` instead of `from`?

Thanks, my mistake.

> Also `u8` would be nicer than `i8` I think. `u8` is the "default" type for a single byte.

I would also expect that but `from_raw_parts_mut` seems to translate both
`*mut c_types::c_void` and `c_types::c_char` to `i8`:

```rust
let data: *mut c_types::c_void = ...;
let buf: &mut [i8] = unsafe { from_raw_parts_mut(data as *mut _, len) };
```
 and that is what I went with. Why that happens then?

Reply:
In C `char` is signed most of the time, so that is what `c_char` is too. In C signed types are often used as they don't require writing `unsigned ...`. In Rust unsigned types are preferred unless it can be negative in practice. The [`Read`](https://doc.rust-lang.org/stable/std/io/trait.Read.html) and [`Write`](https://doc.rust-lang.org/stable/std/io/trait.Write.html) traits for example use `&[u8]`/`&mut [u8]`.

Reply:
I see, thanks for an explanation! Let's stick to `u8` then. When copying bytes, this should be just fine.